pkg/mutation/base_mutator.go

```diff
@@ -0,0 +1,157 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var _ MutatorWithSchema = &BaseMutator{}
+
+func NewMutator(path parser.Path, value string) BaseMutator {
+	return BaseMutator{
+		path:  path,
+		Value: value,
+	}
+}
+
+// Base is a base mutator wrapping the Assign and AssignMetadata type.
+type BaseMutator struct {
+	path  parser.Path
+	Value string
+}
+
+func (m BaseMutator) Mutate(obj *unstructured.Unstructured) error {
+	return mutate(m, obj.Object, nil, 0)
+}
+
+func mutate(m BaseMutator, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(parser.Object).Reference]
+		if !ok {
+			next = createMissingElement(m, current, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		glob := pathEntry.(parser.List).Glob
+		key := pathEntry.(parser.List).KeyField
+		for _, listElement := range current.([]interface{}) {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[key]; ok {
+				if *pathEntry.(parser.List).KeyValue == elementValue {
+					if err := mutate(m, listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !pathEntry.(parser.List).Glob && !elementFound {
+			next := createMissingElement(m, current, previous, depth)
+			if err := mutate(m, next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func addValue(m BaseMutator, current interface{}, previous interface{}, depth int) error {
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		current.(map[string]interface{})[pathEntry.(parser.Object).Reference] = m.Value
+	case parser.ListNode:
+		return addListElementWithValue(m, current, previous, depth)
+	}
+	return nil
+}
+
+func addListElementWithValue(m BaseMutator, current interface{}, previous interface{}, depth int) error {
+	pathEntry := m.Path().Nodes[depth]
+
+	if pathEntry.(parser.List).Glob {
+		return fmt.Errorf("Last path entry can not be globbed")
+	}
+	key := pathEntry.(parser.List).KeyField
+	keyValue := *pathEntry.(parser.List).KeyValue
+
+	for _, listElement := range current.([]interface{}) {
+		if elementValue, ok := listElement.(map[string]interface{})[key]; ok && keyValue == elementValue {
+			return nil // Element is already present, skip the update
+		}
+	}
+	next := make(map[string]interface{})
```

Comment:
this could just be:

```
next  := map[string]interface{}{}
if err := json.Unmarshal([]byte(m.Value), &next); err != nil {
   return err
}
```

would do the job here. I'm assuming that `m.Value` is a string, as is currently the case, though that may not be accurate.

https://github.com/open-policy-agent/gatekeeper/blob/964864d171ffe50bc17ac24c95bde45f4c1f2e0e/apis/mutations/v1alpha1/assign_types.go#L50-L52

Reply:
Neat. 
An error would probably mean it's just a plain value, so does not need decoding.
for example: 
- pathEntry: spec.x, value: {a: z}    
  would result in: spec: {x: {a:z}}
- pathEntry: spec.x, value: test       
  would result in: spec: {x: test}
It would however be difficult to tell if the unmarshaling error means it's a string value (or array?), or if the user made a mistake in the value. Maybe we could a flag or different value in the Assign?

Would it be ok if I open another PR for this change in order to take the discussion there (and get this change in a basic form in quicker)?

Reply:
If the value is a member of an array, it must be an object, since it needs to be able to store the key value.

For object keys, `value`'s type can be left as `interface{}`

I think this is a known property, so shouldn't need much discussion. Since others need to weigh in on this review anyway, it may be worth fixing now?

If not (or I'm missing some subtlety), add a TODO for follow-up so we don't forget

Reply:
Added a TODO for now.

Setting entire object will need some more checks to properly compare existing data to the one unmarshaled.
This would be less of a problem for an object, where we could just raise if the last reference is simply present. For array we would have to compare all existing elements to avoid adding a duplicate to an existing one.